### ISSUE TITLE
Formalise the concept of "singleton" realm type

### ISSUE THREAD

**Issue by ywangd** [elastic/elasticsearch#69958]
There are realm types for which only a single realm can be configured for them. They are informally referred as the "singleton" relam type. For example, the `file`, `native` and `kerberos` realm type all fall into this category (the `reserved` realm also behave this way but it is not configurable so it is less of a concern). Currently, this "singleton" concept is not formalised. Whenever they are needed, ad-hoc checks are performed for it (see [Realms#initRealms](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java#L167) for an example).

It would be useful to formalise this concept into the `Realm` itself, i.e. something that can be actively declared by a `Realm` subclass. It has at least the following benefits:
1. The intention is clear by just look at the each `Realm` subclass as opposed to having to read through the call sites
2. The logic is explicit and can be enforced in one place
3. Custom realms can also declare to be "singleton" and have it checked and enforced the same way as builtin realms.

PS: the word `singleton` is a working term and can possilby change in the actual implementation.

**Comment by bytebilly** [comment]
@ywangd if I get your proposal correctly, this change will not impact on the configuration file that will continue to use the realm name. Is that correct?

**Comment by ywangd** [comment]
@bytebilly Correct. It really is just a refactoring with no user facing changes (unless if you are a developer for custom realms).
